### ISSUE TITLE
allreduce on a subset of processes

### ISSUE THREAD

**Issue by lachlants** [horovod/horovod#105]
Hello, 
Is it possible to call `allreduce` on a subset of Horovod processes? This is useful when your doing distributed batch normalization, i.e. , you estimate the batch statistics for a single node then `allreduce` them across a subset of nodes... this way the batch size used in the batch normalization can be independent from the batch size in the overall gradient computation.

**Comment by alsrgv** [comment]
@lachlants, not at the moment. I think the simplest way to achieve that would be to allocate multiple GPUs per process, do the necessary math and use tower based averaging within the process, and then let Horovod take care of scaling among processes. This is similar to #102.

**Comment by lachlants** [comment]
Is there any major issues with implementing a solution like the one I described? I may have a crack at it since the method you suggest is not feasible with my cluster (I rarely get assigned two GPUs on the same machine) and batch normalization batch-size seems crucial to my models.

**Comment by alsrgv** [comment]
I think having an `allreduce`, `broadcast` or `allgather` on a set of specific ranks along with an example use case like yours would be a good addition.

**Comment by lachlants** [comment]
Actually just noticed the `allgather` method, I should be able to implement the batch norm averaging with it...  won't be as efficent as a ring based `allreduce`, but shouldn't really matter since the batch norm statistics are relatively tiny compared to the gradients.

**Comment by alsrgv** [comment]
@lachlants, that's a good idea! Do you mind sharing an example code when you make it work?

**Comment by lkmokadam** [comment]
I think you can try [this](https://github.com/lkmokadam/MultiCommHorovod.git). It can allow you to train model on mutually exclusive set of nodes. 

On calling hvd.init([[0,1],[2,3]]), it will create two communicators will be created for communications, one with rank 0 and 1, another with rank 2,3.

If you find useful, I can create PR from it.

**Comment by alsrgv** [comment]
@lkmokadam, if I understand the code correctly, if you run the multicomm example on 4 GPUs, you'd get two MNIST models trained on 2 GPUs each?  This may be useful in #216, @mkovalski could specify all his actual allreduce ranks in the first group, and all the preprocessing ranks in the second group (and never use them).

I am wondering though if it will be useful to be able to specify different groups for different operations.  E.g. you could have operations that do allreduce within node, and other operations which would do allreduce across nodes.  In this model, we'd still have one global communicator that would do orchestration, but then for actual operations like ncclAllReduce or MPI_Allreduce we could dynamically create communicators on the fly (and cache them) for any group configuration that was used.

API-wise, it would look like `hvd.allreduce(tensor, ranks=[0, 1])`

What do you guys think?

**Comment by lkmokadam** [comment]
API-wise this will be great addition. But, for better performance, we have to consider various aspects. Most basic is what is maximum number of processes do we have to support?  This will be crucial for deciding communication pattern. All further questions depend on this. 

Considering the communication pattern of the operation that do allreduce within node, and other operations which would do allreduce across nodes, I think this is already handled by MPI using various heuristics. As far as I know, there is no doc available for this. You can get some details [here](https://github.com/open-mpi/ompi/tree/master/ompi/mca/coll/tuned). Performance tests required to backup this.  As NCCL2 is not open source, there is no way to know how NCCL2 perform internode collective operations. But I think they are using some sort of heuristics to perform collective operations because NVIDIA claims NCCL2 provides **multi-GPU collective** communication primitives that are **topology-aware**.

**Comment by alsrgv** [comment]
I believe both MPI and NCCL2 can do the right thing here.

Let me give a more concrete example how I think specifying ranks per operation may be helpful in a context of batch normalization.

One of the problems with object detection models is that they consume so much GPU memory that you have to put only 1 or 2 examples per GPU.  With that, it's impossible to run batch normalization in training mode, because you will get very bad results.  At the same time, normalizing batches across all GPUs lead to too large batch size and produces bad results, too.

With having a per-operation set of ranks, we could do forward pass batch normalization which would normalize inputs across a subset of ranks, e.g. if you have 4 DGX-1 servers, you could normalize batches across ranks `[0..15]` and `[16..31]`.  Gradient allreduce during the backward pass, however, would happen across all ranks.

This behavior cannot be achieved with ranks being passed to `hvd.init()`.  So, I think having ranks passed into individual ops is more powerful.

To make certain things simpler in situations where the script always uses the same set of ranks, we could still allow `hvd.init(default_ranks=[1, 3, 5])` which would be equivalent of always specifying `hvd.<op>(..., ranks=[1, 3, 5])`, but that I think that setting should not modify the value of `hvd.rank()` - it would still be equal to `rank` of `COMM_WORLD`.  Additionally, different workers could call `hvd.init()` with different `default_ranks`.

Would this approach work for your scenarios of multi-communication?

Implementation & performance-wise, this would still use MPI & NCCL group communicators for data-intensive operations.  The only communication happening via `COMM_WORLD` would be MPI orchestration messages which are very small and batched - I believe it should scale to tens of thousands of processes.  We can figure out some hierarchical orchestration if it doesn't.

**Comment by alsrgv** [comment]
@mkovalski, @lachlants, do you have any opinions about being able to specify ranks per operation, e.g. via `hvd.allreduce(..., ranks=[1, 3, 5])`, and being able to specify a default set of ranks via `hvd.init()`, e.g. `hvd.init(default_ranks=[1, 3, 5])`?  `hvd.rank()` would still return an actual `COMM_WORLD` rank.  We can shield other MPI communication from tag pollution by creating a private communicator that will mirror `COMM_WORLD`.

**Comment by mkovalski** [comment]
Agreed with setting the ranks in hvd.init() should not change any of the ranks or size of the communicator, it should mirror however many processes you ran it with via MPI. IE running mpirun -n 4 and only using 2 of those procs for the allreduce method at training time, hvd.size() should still give you 4. I think this gives the user more control over what they can do with the remaining procs not used during training.

**Comment by alsrgv** [comment]
Excellent.  Then it's just a small matter of implementing it :-)

**Comment by sarunyap** [comment]
Hi @alsrgv, are you currently working on implementing this functionality?

Thanks!

**Comment by maxhgerlach** [comment]
@alsrgv Is this functionality still planned for Horovod? It would be very helpful for several of our use cases.

What would you estimate: How much work would it be to implement this without prior knowledge of the code base?

**Comment by alsrgv** [comment]
@maxhgerlach, one thing that can be done today is `hvd.init(comm=[list-of-ranks])` (see [here](https://github.com/horovod/horovod/blob/master/horovod/common/__init__.py#L58)), so you could pre-partition your MPI job into multiple small Horovod jobs.  But we don't have support for on-demand subgroup allreduce where you could specify the list of ranks per operation yet.

It's a pretty big change, as we'll need to make both coordinator and all communication ops aware of subgroup concept.  The code is concentrated in [operations.cc](https://github.com/horovod/horovod/blob/master/horovod/common/operations.cc) and [ops](https://github.com/horovod/horovod/tree/master/horovod/common/ops).  You can browse through and decide if it looks scary or not, and if it doesn't - a contribution would be very welcome :-)

**Comment by maxhgerlach** [comment]
@alsrgv thanks for the pointers! At the moment I fear priorities at work make it hard to commit to the effort, although I'd find it quite interesting.

**Comment by andfoy** [comment]
> It's a pretty big change, as we'll need to make both coordinator and all communication ops aware of subgroup concept. The code is concentrated in operations.cc and ops. You can browse through and decide if it looks scary or not, and if it doesn't - a contribution would be very welcome :-)

Hi @alsrgv, do you have any hints to get working on this one? I'm very interested on implementing such functionality

**Comment by weidai0903** [comment]
@alsrgv I have a question that is similar to this issue. We are trying to implement a functionality that makes multiple subgroups run in parallel. I have already make one subgroup work without problem, but when it comes to multiple subgroups, there is some hard problem with the response cache. For instance, in the controller,  when it tries to update the cache, it considers all tensors on all ranks. However in the case of multiple subgroups, certain tensor might belong to different subgroup aka communication group. What would be a clean approach to this matter?
1. Do you think create different controllers for different subgroup(communication group) would be a good idea for the purpose above?
2. Do you think create multiple backgroundLoop thread or each subgroup would be a better idea?
Any advice? Thanks.

**Comment by jarednielsen** [comment]
Running allreduce on a subset of ranks would unblock my use case of language models. I pretrain a language models on N nodes, 8*N ranks. However, for finetuning periodically throughout the pretraining process (for intermediate evaluation on downstream tasks), I only use 1 node and 8 ranks. A barrier op like `hvd.allreduce(tf.random.normal([1]))` doesn't work for the unutilized nodes because it conflicts with the true allreduce that the finetuning ranks are performing.

**Comment by maxhgerlach** [comment]
There now is an issue #2139 discussing this potential enhancement.
